File: Numeros Pseudo-Aleatorios/Blum Blum Shub/BlumBlumShub.go
/* 
 * PRNG .......: Blum-Blum-Shub
 * Ano ........: 1986
 * Autor ......: L. Blum, M. Blum & M. Shub
*/

package main

import (
	"fmt"
	"time"
)

func BlumBlumShub(X *uint32) uint32 {
	*X = (*X * *X) % 2467246241
	return *X
}

var Xn1 [10]uint32
var Xn2 [10]uint32
var X1, X2 uint32

func main() {
	
	fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
	fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
	fmt.Println("PADRAO: Blum-Blum-Shub")
	fmt.Println("==========================================\n")
	fmt.Println("    Semente: 1234        Semente: TIME")
	fmt.Println("    -------------        -------------\n")

	X1 = 1234
	X2 = uint32(time.Now().Unix())

	for i := 0; i <= 9; i++ {
		Xn1[i] = BlumBlumShub(&X1)
		Xn2[i] = BlumBlumShub(&X2)
		fmt.Printf("%17d        %13d\n", Xn1[i], Xn2[i])
	}
	
}

File: Numeros Pseudo-Aleatorios/Blum Blum Shub/BlumBlumShub.pas
{ 
 PRNG .......: Blum-Blum-Shub
 Ano ........: 1986
 Autor ......: L. Blum, M. Blum & M. Shub
}

program BlumBlumShub;

uses
  SysUtils, DateUtils;
  
function BlumBlumShub(var X: LongWord): LongWord;
begin
  X := (X * X) mod 2467246241;
  BlumBlumShub := X;
end;

var
  Xn1: array[1..10] of LongWord;
  Xn2: array[1..10] of LongWord;
  X1, X2: LongWord;
  i: Integer;
  
begin

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Blum-Blum-Shub');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  X1 := 1234;
  X2 := UInt32(DateTimeToUnix(Now));
  
  for i := 1 to 10 do
  begin
    Xn1[i] := BlumBlumShub(X1);
    Xn2[i] := BlumBlumShub(X2);
    writeln(Xn1[i]:17, '        ', Xn2[i]:13);
  end

end.
